### PR TITLE
Don't crash when return type jsdoc tag is malformed

### LINKED ISSUE #6662: Salsa does not show doc comment when I have "@returns" in it?
_From @alexandrudima on January 27, 2016 11:1_

Testing #2218 

For the following JS code I get no hover on `a1()`:

```javascript
/**
 * This is a very cool function that is very nice.
 * @returns something
 */
function a1() {
        try {
                        throw new Error('x');         } catch(x) { x--; }
                        return 23;
}

x - a1()
```

If I remove the `@returns` from the js doc comment, then I get a hover, i.e. it works for this:

```javascript
/**
 * This is a very cool function that is very nice.
 */
function a1() {
        try {
                        throw new Error('x');         } catch(x) { x--; }
                        return 23;
}

x - a1()

```

_Copied from original issue: Microsoft/vscode#2436_

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes #6662

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -10344,7 +10344,7 @@ namespace ts {
 
         function getReturnTypeFromJSDocComment(func: SignatureDeclaration | FunctionDeclaration): Type {
             const returnTag = getJSDocReturnTag(func);
-            if (returnTag) {
+            if (returnTag && returnTag.typeExpression) {
                 return getTypeFromTypeNode(returnTag.typeExpression.type);
             }
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Can you explicitly return `undefined`?

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Or should you be returning the `unknownType` instead?

File: tests/cases/fourslash/getJavaScriptQuickInfo7.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+//// 	try {
+//// 		throw new Error('x');
+//// 	} catch (x) { x--; }
+//// 	return 23;
+//// }
+//// 
+//// x - /**/a1()
+
+goTo.marker();
+verify.quickInfoExists();
```
Comment (billti) [inline tests/cases/fourslash/getJavaScriptQuickInfo7.ts]: Can we also verify that the quickinfo still correctly contains the first line of the comment?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -10344,9 +10344,11 @@ namespace ts {
 
         function getReturnTypeFromJSDocComment(func: SignatureDeclaration | FunctionDeclaration): Type {
             const returnTag = getJSDocReturnTag(func);
-            if (returnTag) {
+            if (returnTag && returnTag.typeExpression) {
```
Comment (billti) [inline src/compiler/checker.ts]: Do we have tests for malformed types in other tags (e.g. `@param`)?

**Comment by billti** [comment]
Couple of minor comments, else :+1:

File: src/compiler/checker.ts (hunk 1 above)
Comment (mhegazy) [inline src/compiler/checker.ts]: undefined is fine.

**Comment by mhegazy** [comment]
:+1:

File: tests/cases/fourslash/getJavaScriptQuickInfo7.ts (hunk 2 above)
Comment (RyanCavanaugh) [inline tests/cases/fourslash/getJavaScriptQuickInfo7.ts]: Verified manually since Fourslash is currently really bad at that

**Pull request merged**